{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Autotrain 을 가지고 Llama2 데이터 학습 : PEFT\n",
    "    * Autotrain 을 로컬데이터를 학습할 수 있는 방법을 찾지 못함.\n",
    "    * 허깅페이스의 자원을 이용해서 모델학습을 시켜야 함.\n",
    "```\n",
    "1. excute: https://hackernoon.com/ko/%EC%A7%91%EC%97%90%EC%84%9C-%EB%A7%8C%EB%93%A0-%EB%8C%80%ED%98%95-%EC%96%B8%EC%96%B4-%EB%AA%A8%EB%8D%B8%EC%9D%84-%EC%8B%A4%ED%97%98%ED%95%98%EB%8A%94-%EC%95%8C%ED%8C%8C%EC%B9%B4-%EB%A1%9C%EB%9D%BC\n",
    "2. https://github.com/tloen/alpaca-lora/blob/main/finetune.py\n",
    "3. model: Llama2 pytouch.bin\n",
    "4. parameter data type: touch.float16\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### requiment modules\n",
    "```\n",
    "pip install autotrain-advanced\n",
    "pip install transformers\n",
    "pip install bitsandbytes\n",
    "pip install accelerate\n",
    "pip install sentencepiece\n",
    "pip install protobuf\n",
    "pip install xformers\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/bwis/aibiseo/ipynb/4-train/../script/train/run_clm.py\", line 637, in <module>\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/bwis/aibiseo/ipynb/4-train/../script/train/run_clm.py\", line 637, in <module>\n",
      "    main()\n",
      "  File \"/home/bwis/aibiseo/ipynb/4-train/../script/train/run_clm.py\", line 239, in main\n",
      "    main()\n",
      "  File \"/home/bwis/aibiseo/ipynb/4-train/../script/train/run_clm.py\", line 239, in main\n",
      "    model_args, data_args, training_args = parser.parse_args_into_dataclasses()\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/hf_argparser.py\", line 338, in parse_args_into_dataclasses\n",
      "    model_args, data_args, training_args = parser.parse_args_into_dataclasses()\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/hf_argparser.py\", line 338, in parse_args_into_dataclasses\n",
      "    obj = dtype(**inputs)\n",
      "      File \"<string>\", line 123, in __init__\n",
      "obj = dtype(**inputs)\n",
      "  File \"<string>\", line 123, in __init__\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/training_args.py\", line 1528, in __post_init__\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/training_args.py\", line 1528, in __post_init__\n",
      "    and (self.device.type != \"cuda\")\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/training_args.py\", line 1995, in device\n",
      "    return self._setup_devices\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/utils/generic.py\", line 56, in __get__\n",
      "    cached = self.fget(obj)\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/training_args.py\", line 1927, in _setup_devices\n",
      "    and (self.device.type != \"cuda\")\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/training_args.py\", line 1995, in device\n",
      "    self.distributed_state = PartialState(timeout=timedelta(seconds=self.ddp_timeout))\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/accelerate/state.py\", line 171, in __init__\n",
      "    return self._setup_devices\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/utils/generic.py\", line 56, in __get__\n",
      "    assert (\n",
      "AssertionError: DeepSpeed is not available => install it using `pip3 install deepspeed` or build it from source\n",
      "    cached = self.fget(obj)\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/transformers/training_args.py\", line 1927, in _setup_devices\n",
      "    self.distributed_state = PartialState(timeout=timedelta(seconds=self.ddp_timeout))\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/accelerate/state.py\", line 171, in __init__\n",
      "    assert (\n",
      "AssertionError: DeepSpeed is not available => install it using `pip3 install deepspeed` or build it from source\n",
      "[2024-03-13 19:57:50,512] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 0 (pid: 418572) of binary: /data/pvenv/train/bin/python\n",
      "Traceback (most recent call last):\n",
      "  File \"/data/pvenv/train/bin/torchrun\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py\", line 347, in wrapper\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/torch/distributed/run.py\", line 812, in main\n",
      "    run(args)\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/torch/distributed/run.py\", line 803, in run\n",
      "    elastic_launch(\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/torch/distributed/launcher/api.py\", line 135, in __call__\n",
      "    return launch_agent(self._config, self._entrypoint, list(args))\n",
      "  File \"/data/pvenv/train/lib/python3.10/site-packages/torch/distributed/launcher/api.py\", line 268, in launch_agent\n",
      "    raise ChildFailedError(\n",
      "torch.distributed.elastic.multiprocessing.errors.ChildFailedError: \n",
      "============================================================\n",
      "../script/train/run_clm.py FAILED\n",
      "------------------------------------------------------------\n",
      "Failures:\n",
      "[1]:\n",
      "  time      : 2024-03-13_19:57:50\n",
      "  host      : gpu-1.novalocal\n",
      "  rank      : 1 (local_rank: 1)\n",
      "  exitcode  : 1 (pid: 418573)\n",
      "  error_file: <N/A>\n",
      "  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html\n",
      "------------------------------------------------------------\n",
      "Root Cause (first observed failure):\n",
      "[0]:\n",
      "  time      : 2024-03-13_19:57:50\n",
      "  host      : gpu-1.novalocal\n",
      "  rank      : 0 (local_rank: 0)\n",
      "  exitcode  : 1 (pid: 418572)\n",
      "  error_file: <N/A>\n",
      "  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "!torchrun --nproc_per_node=2 --master_port=34321 ../script/train/run_clm.py \\\n",
    "--model_name_or_path='/data/aibiseo/models/casllm-base-7b-hf' \\\n",
    "--train_file='/data/aibiseo/dataset/en_alpaca_data.json' \\\n",
    "--num_train_epochs=2 \\\n",
    "--block_size=1024 \\\n",
    "--per_device_train_batch_size=1 \\\n",
    "--gradient_accumulation_steps=64 \\\n",
    "--torch_dtype=float16 \\\n",
    "--fp16 \\\n",
    "--output_dir='/data/aibiseo/models/casalpha-sllm-7bhf' \\\n",
    "--deepspeed=ds_zero3-nooffload.json \\\n",
    "--do_train \\\n",
    "--save_strategy='epoch' \\\n",
    "--logging_strategy='steps' \\\n",
    "--logging_first_step \\\n",
    "--save_total_limit=1 \\\n",
    "--run_name='casalpha-sllm-7bhf-v1.1b'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "train",
   "language": "python",
   "name": "train"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
